### ISSUE TITLE
about the confidence score of keypoints

### ISSUE THREAD

**Issue by xiaoxian-m** [mks0601/3DMPPE_POSENET_RELEASE#47]
hi, it is really a great work!

But I have a question about the output. It is clearly the keypoints coordinates can be got from the volumetric heatmaps, but if I can got the confidence score about the keypoint? I try to got them by take the max value like the 2d heatmap, but it seems not convincing. Could you give me some advice?

**Comment by mks0601** [comment]
Hi, sorry for the late reply. Why did you use 2D heatmap? You can use the 3D heatmap, which is the output o the PoseNet.

**Comment by xiaoxian-m** [comment]
> Hi, sorry for the late reply. Why did you use 2D heatmap? You can use the 3D heatmap, which is the output o the PoseNet.

yes, it workes for the coordinates, but I want to know if I can get the pnt confidence score like the 2d heatmaps?

Look forward your relpy, thank you.

**Comment by mks0601** [comment]
You can perform agrmax on 3d heatmap

**Comment by xiaoxian-m** [comment]
yeah, I tried got score like follows:
```
visible = nn.Sequential(nn.AdaptiveMaxPool2d(1))
    score = visible(heatmaps.reshape(-1, joint_num, cfg.depth_dim, cfg.output_shape[0]*cfg.output_shape[1])).reshape(-1, joint_num, 1)
```

But I got the reults like this:

```
tensor([[[35.6810, 28.9530, 32.0002,  0.0548],
         [39.0421, 28.5499, 33.5959,  0.0362],
         [38.6442, 41.4325, 33.0469,  0.0121],
         [40.2233, 53.6736, 28.7742,  0.1204],
         [32.3229, 29.4012, 30.2769,  0.0360],
         [30.0288, 40.1131, 38.6354,  0.0213],
         [29.2661, 50.7853, 45.3092,  0.0435],
         [36.0431, 22.1684, 29.8909,  0.0151],
         [36.0950, 15.2892, 31.3522,  0.0064],
         [34.3501, 13.7839, 34.0876,  0.0211],
         [34.8420, 10.7147, 33.4273,  0.0590],
         [32.5901, 15.2547, 29.2909,  0.0078],
         [26.3157, 15.1706, 27.8022,  0.0230],
         [22.3973, 18.7131, 33.3000,  0.1120],
         [39.5775, 17.5060, 33.0131,  0.0110],
         [41.6239, 24.3828, 36.3712,  0.0467],
         [35.8125, 25.5394, 40.2284,  0.0642],
         [36.1411, 16.5415, 31.0699,  0.0059]]], device='cuda:0')
```
which vis result like this:
![image](https://user-images.githubusercontent.com/33691506/80695483-8ffdde00-8b08-11ea-8de4-b51bfc55f384.png)

the forth column is the points score, I dont know if I do something wrong, I think the score should be range in 0 and 1.

**Comment by mks0601** [comment]
Try this.
```
heatmaps = F.softmax(heatmaps, 2) # shape: (batch_size, joint_num, cfg.depth_dim*cfg.output_shape[0]*cfg.output_shape[1])
score, _ = torch.max(heatmaps, 2) # shape: (batch_size, joint_num)
```

**Comment by xiaoxian-m** [comment]
> score, _ = torch.max(heatmaps, 2) # shape: (batch_size, joint_num)

I tried the method and got the result like this:

```
tensor([[[35.6810, 28.9530, 32.0002,  0.0548],
         [39.0421, 28.5499, 33.5959,  0.0362],
         [38.6442, 41.4325, 33.0469,  0.0121],
         [40.2233, 53.6736, 28.7742,  0.1204],
         [32.3229, 29.4012, 30.2769,  0.0360],
         [30.0288, 40.1131, 38.6354,  0.0213],
         [29.2661, 50.7853, 45.3092,  0.0435],
         [36.0431, 22.1684, 29.8909,  0.0151],
         [36.0950, 15.2892, 31.3522,  0.0064],
         [34.3501, 13.7839, 34.0876,  0.0211],
         [34.8420, 10.7147, 33.4273,  0.0590],
         [32.5901, 15.2547, 29.2909,  0.0078],
         [26.3157, 15.1706, 27.8022,  0.0230],
         [22.3973, 18.7131, 33.3000,  0.1120],
         [39.5775, 17.5060, 33.0131,  0.0110],
         [41.6239, 24.3828, 36.3712,  0.0467],
         [35.8125, 25.5394, 40.2284,  0.0642],
         [36.1411, 16.5415, 31.0699,  0.0059]]], device='cuda:0')
```

Is that right?

**Comment by mks0601** [comment]
Did you check the shape of the `heatmaps`? You should do this before running the codes.
```
heatmaps = heatmaps.view(batch_size, joint_num, cfg.depth_dim * cfg.output_shape[0] * cfg.output_shape[1])
```

**Comment by xiaoxian-m** [comment]
yes, the whole code I use as bellows:
```
def soft_argmax(heatmaps, joint_num):

    **heatmaps = heatmaps.reshape((-1, joint_num, cfg.depth_dim*cfg.output_shape[0]*cfg.output_shape[1]))
    heatmaps = F.softmax(heatmaps, 2)

    score, _ = torch.max(heatmaps, 2) # shape: (batch_size, joint_num)**
    heatmaps = heatmaps.reshape((-1, joint_num, cfg.depth_dim, cfg.output_shape[0], cfg.output_shape[1]))

    accu_x = heatmaps.sum(dim=(2,3))
    accu_y = heatmaps.sum(dim=(2,4))
    accu_z = heatmaps.sum(dim=(3,4))

    accu_x = accu_x * torch.cuda.comm.broadcast(torch.arange(1,cfg.output_shape[1]+1).type(torch.cuda.FloatTensor), devices=[accu_x.device.index])[0]
    accu_y = accu_y * torch.cuda.comm.broadcast(torch.arange(1,cfg.output_shape[0]+1).type(torch.cuda.FloatTensor), devices=[accu_y.device.index])[0]
    accu_z = accu_z * torch.cuda.comm.broadcast(torch.arange(1,cfg.depth_dim+1).type(torch.cuda.FloatTensor), devices=[accu_z.device.index])[0]

    accu_x = accu_x.sum(dim=2, keepdim=True) -1
    accu_y = accu_y.sum(dim=2, keepdim=True) -1
    accu_z = accu_z.sum(dim=2, keepdim=True) -1

    **score = score.reshape(-1,joint_num, 1)
    coord_out = torch.cat((accu_x, accu_y, accu_z, score), dim=2)**

    return coord_out
```

Anything wrong?

**Comment by mks0601** [comment]
I see. You are printing coordinates as well. I thought you are printing only scores but there are some numbers greater than 1, which is weird. Anyway, the score can be small, especially it is from softmax function over 64*64*64 dimension.
